[Executorch][Recipes][Coreml] Add coreml backend recipes

abhinaykukkadapu opened the pull request:
### Summary
Adds coreml recipes as discussed similar to xnnpack backend recipes to use them for export flow.

Fixes #13100

### Test plan
`python -m unittest backends/apple/coreml/test/test_coreml_recipes.py`

metascroy commented on backends/apple/coreml/recipes/coreml_recipe_types.py:
```diff
@@ -0,0 +1,32 @@
+COREML_BACKEND: str = "coreml"
+
+
+class CoreMLRecipeType(RecipeType):
```
@abhinaykukkadapu I don't think precision/compute unit lead to different recipe types any more than the deployment target does (which is given as a separate arg).

The reason they're in the name in optimum-et is because they had to be strings to work with the CLI:

```
optimum-cli export model_id --recipe coreml_gpu_f32
```

But the ET recipes are a bit more general, so maybe they could be args here as well?

abhinaykukkadapu replied:
Sounds good, will update the in the next rev.

metascroy commented on backends/apple/coreml/recipes/coreml_recipe_types.py:
```diff
@@ -0,0 +1,32 @@
+    # FP32 precision with different compute units
+    FP32_CPU = "coreml_fp32_cpu"
+    FP32_GPU = "coreml_fp32_gpu"
+    FP32_NEURAL_ENGINE = "coreml_fp32_neural_engine"
```
Nit: the neural engine doesn't support FP32.  Maybe call it NE instead of neaural_engine to align with coremltools naming convention?

kimishpatel replied:
For canonical recipes we probably just want to go with FP16. This way you dont have combinatorial explosion either. If we want to construct a reciped then provide options to chose cu,  precision etc.

metascroy commented on backends/apple/coreml/test/test_coreml_recipes.py:
```diff
@@ -0,0 +1,207 @@
+from torch import nn
+from torch.testing._internal.common_quantization import TestHelperModules
+
+try:
```
When is coremltools not available?

It's a required dependency in ET now

metascroy commented on backends/apple/coreml/recipes/coreml_recipe_provider.py:
```diff
@@ -0,0 +1,156 @@
+
+        return ExportRecipe(
+            name=recipe_type.value,
+            quantization_recipe=None,  # TODO - add quantization recipe
```
FYI: I have a PR updating CoreML docs that you might check out for quantization and available partitioner options: https://github.com/pytorch/executorch/pull/13120

kimishpatel replied:
@metascroy it might make sense for you to follow up on the recipes for quant

abhinaykukkadapu replied:
@metascroy If you want to follow up, i can leave it upto you as you have more context on quantization, or i can take it up if you don't have bandwidth.

abhinaykukkadapu replied:
@metascroy sure will add in a follow up diff, can we go ahead with these changes?

metascroy commented on backends/apple/coreml/recipes/coreml_recipe_provider.py:
```diff
@@ -0,0 +1,156 @@
+
+        edge_compile_config = EdgeCompileConfig(
+            _check_ir_validity=False,
+            _skip_dim_order=True,
```
CoreML now supports dim_order

abhinaykukkadapu committed:
Add coreml backend recipes (#13121)

Summary:
Adds coreml recipes as discussed similar to xnnpack backend recipes to use them for export flow.

Fixes https://github.com/pytorch/executorch/issues/13100

Pull Request resolved: https://github.com/pytorch/executorch/pull/13121

Test Plan: `python -m unittest backends/apple/coreml/test/test_coreml_recipes.py`

Reviewed By: metascroy

Differential Revision: D79654041

Pulled By: abhinaykukkadapu
